pkg/cluster/volumes.go

```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		return fmt.Errorf("could not list persistent volumes: %v", err)
+	}
+	for _, pv := range pvs {
+		for _, resizer := range resizers {
+			if !resizer.VolumeBelongsToProvider(pv) {
+				continue
+			}
+			totalCompatible += 1
+			if !resizer.IsConnectedToProvider() {
+				err := resizer.ConnectToProvider()
+				if err != nil {
+					return fmt.Errorf("could not connect to the volume provider: %v", err)
+				}
+				defer resizer.DisconnectFromProvider()
+			}
+			volumeSize := quantityToGigabyte(pv.Spec.Capacity[v1.ResourceStorage])
+			if volumeSize > newSize {
+				return fmt.Errorf("cannot shrink persistent volume")
+			}
+			if volumeSize != newSize {
+				awsVolumeId, err := resizer.GetProviderVolumeID(pv)
+				if err != nil {
+					return err
+				}
+				c.logger.Debugf("updating persistent volume %s to %d", pv.Name, newSize)
+				if err := resizer.ResizeVolume(awsVolumeId, newSize); err != nil {
+					return fmt.Errorf("could not resize EBS volume %s: %v", awsVolumeId, err)
+				}
+				c.logger.Debugf("resizing the filesystem on the volume %s", pv.Name)
+				podName := getPodNameFromPersistentVolume(pv)
+				if err := c.resizePostgresFilesystem(podName, []filesystems.FilesystemResizer{&filesystems.Ext234Resize{}}); err != nil {
+					return fmt.Errorf("could not resize the filesystem on pod '%s': %v", podName, err)
+				}
+				c.logger.Debugf("filesystem resize successfull on volume %s", pv.Name)
+				pv.Spec.Capacity[v1.ResourceStorage] = newQuantity
+				c.logger.Debugf("updating persistent volume definition for volume %s", pv.Name)
+				if _, err := c.KubeClient.PersistentVolumes().Update(pv); err != nil {
+					return fmt.Errorf("could not update persistent volume: %s", err)
+				}
+				c.logger.Debugf("successfully updated persistent volume %s", pv.Name)
+			}
+		}
+	}
+	if len(pvs) > 0 && totalCompatible == 0 {
+		return fmt.Errorf("could not resize EBS volumes: persistent volumes are not compatible with existing resizing providers")
+	}
+	return nil
+}
+
+func (c *Cluster) VolumesNeedResizing(newVolume spec.Volume) (bool, error) {
+	volumes, manifestSize, err := c.listVolumesWitManifestSize(newVolume)
+	if err != nil {
+		return false, err
+	}
+	for _, pv := range volumes {
+		currentSize := quantityToGigabyte(pv.Spec.Capacity[v1.ResourceStorage])
+		if currentSize != manifestSize {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (c *Cluster) listVolumesWitManifestSize(newVolume spec.Volume) ([]*v1.PersistentVolume, int64, error) {
+	newSize, err := resource.ParseQuantity(newVolume.Size)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not parse volume size from the manifest: %v", err)
+	}
+	manifestSize := quantityToGigabyte(newSize)
+	volumes, err := c.listPersistentVolumes()
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not list persistent volumes: %v", err)
+	}
+	return volumes, manifestSize, nil
+}
+
+func getPodNameFromPersistentVolume(pv *v1.PersistentVolume) *spec.NamespacedName {
```

Comment:
Wouldn't it be confusing that we return not a claim, but a pod name, as the code does `pv.Spec.ClaimRef.Name[len(constants.DataVolumeName)+1:]`?